diff --git a/modules/setting/setting.go b/modules/setting/setting.go
index 7be0842..f3d01dd 100644
--- a/modules/setting/setting.go
+++ b/modules/setting/setting.go
@@ -778,7 +778,7 @@ func loadFromConf(allowEmpty bool, extraConfig string) {
 	OfflineMode = sec.Key("OFFLINE_MODE").MustBool()
 	DisableRouterLog = sec.Key("DISABLE_ROUTER_LOG").MustBool()
 	if len(StaticRootPath) == 0 {
-		StaticRootPath = AppWorkPath
+		StaticRootPath = "@data@"
 	}
 	StaticRootPath = sec.Key("STATIC_ROOT_PATH").MustString(StaticRootPath)
 	StaticCacheTime = sec.Key("STATIC_CACHE_TIME").MustDuration(6 * time.Hour)
