diff --git a/gen-wcwidth.py b/gen-wcwidth.py
index 09f4b9a6..a755d77c 100755
--- a/gen-wcwidth.py
+++ b/gen-wcwidth.py
@@ -224,6 +224,9 @@ def parse_emoji() -> None:
 doublewidth: Set[int] = set()
 ambiguous: Set[int] = set()
 
+doublewidth |= set(range(0xF1900, 0xF1990)) # nimi pu and nimi ku suli
+doublewidth |= set(range(0xf19A0, 0xF1a00)) # nimi ku lili
+
 
 def parse_eaw() -> None:
     global doublewidth, ambiguous
diff --git a/kitty/unicode-data.h b/kitty/unicode-data.h
index ba4f4d40..c945a293 100644
--- a/kitty/unicode-data.h
+++ b/kitty/unicode-data.h
@@ -37,7 +37,7 @@ can_strip_from_end_of_url(uint32_t ch) {
 
 static inline bool
 is_private_use(char_type ch) {
-    return (0xe000 <= ch && ch <= 0xf8ff) || (0xF0000 <= ch && ch <= 0xFFFFF) || (0x100000 <= ch && ch <= 0x10FFFF);
+    return ((0xe000 <= ch && ch <= 0xf8ff) || (0xF0000 <= ch && ch <= 0xFFFFF) || (0x100000 <= ch && ch <= 0x10FFFF)) && !(0xF1900 >= ch && ch <= 0xF19FF);
 }
 
 
