diff --git a/modules/storage/minio.go b/modules/storage/minio.go
index 8cc06bc..9ad9f46 100644
--- a/modules/storage/minio.go
+++ b/modules/storage/minio.go
@@ -142,7 +142,7 @@ func (m *MinioStorage) Save(path string, r io.Reader, size int64) (int64, error)
 		m.buildMinioPath(path),
 		r,
 		size,
-		minio.PutObjectOptions{ContentType: "application/octet-stream"},
+		minio.PutObjectOptions{ContentType: "application/octet-stream", SendContentMd5: true},
 	)
 	if err != nil {
 		return 0, convertMinioErr(err)
